In [1]:
# Import libraries
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    f1_score,
    precision_score,
    recall_score,
)
from sklearn.model_selection import train_test_split

# load the optdigits datset directly from sklearn 
digits = load_digits()
X= digits.data
y = digits.target 

print("Dataset loaded successfully")
print(f"Features shape: {X.shape}, Labels shape: {y.shape}")

Dataset loaded successfully
Features shape: (1797, 64), Labels shape: (1797,)


# splitting the data into training and test sets. 
The purpose of splitting the data is to evaluate our model's true performance. 
the trainig set (X_train, y_train) is used by the algorithym to learn underlying patterns. Whereas the test set (X_test, y_test) acts as acompletely unseen evaluation benchmark. Keep them separate ensure the model can generalize to new data instead of just memorizing the training set. 

In [2]:
# Split the data into training (80%) and testing (20%) sets. 
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)



# Parameter selection and rationale.
Parameter to tune: n_estimators
Rationale for choosing this parameter:
'n_estimators' controls the number of decision trees in the random forest ensemble. Increasing this value generally makes the model more robust and stable by averaging out variance across more trees, which reduces the risk of overfitting without introducing bias.

Selection of value for testing: 150
Rationale for choice:
The default in older versions of scikit-learn was 10, while newer versions default to 100. A value of 150 provides a dense enough forest to achieve high stability on a relatively small dataset like optdigits (1,797 samples total) without bottlenecking computation speed.

# Initialize and train model. 

In [5]:
# Create the RandomForestClassifier with our chosen parameter value 
clf = RandomForestClassifier(n_estimators=150, random_state=42)
print("Classifier Settings: ")
print(clf)

print("\nTraining the model... (this might take a few minutes)")
# train the model
clf.fit(X_train, y_train)

print("Training complete! Making predictions..")
# make predictions on the test set. 
preds = clf.predict(X_test)

Classifier Settings: 
RandomForestClassifier(n_estimators=150, random_state=42)

Training the model... (this might take a few minutes)
Training complete! Making predictions..


# Confusion Matrix and Misclassification Analysis. 

In [6]:
# Generate and display the confusion matrix
cm = confusion_matrix(y_test, preds)
print("Confusion Matrix:")
print(cm)

# Analyze misclassifications programmatically
# Sum rows to find totals, subtract the diagonal (correct answers) to find errors per class
errors_per_class = cm.sum(axis=1) - cm.diagonal()
highest_error_class = errors_per_class.argmax()

print("\n--- Misclassification Analysis ---")
if errors_per_class[highest_error_class] == 0:
    print("Amazing! The model made zero misclassifications on the test set.")
else:
    print(
        f"The class with the highest number of misclassifications is: Class {highest_error_class}"
    )
    print(
        f"Total mistakes for this class: {errors_per_class[highest_error_class]}"
    )

Confusion Matrix:
[[32  0  0  0  1  0  0  0  0  0]
 [ 0 28  0  0  0  0  0  0  0  0]
 [ 0  0 33  0  0  0  0  0  0  0]
 [ 0  0  0 32  0  1  0  0  1  0]
 [ 0  0  0  0 46  0  0  0  0  0]
 [ 0  0  0  0  0 45  1  0  0  1]
 [ 0  0  0  0  0  1 34  0  0  0]
 [ 0  0  0  0  0  0  0 33  0  1]
 [ 0  2  0  0  0  0  0  0 28  0]
 [ 0  0  0  0  0  1  0  1  0 38]]

--- Misclassification Analysis ---
The class with the highest number of misclassifications is: Class 3
Total mistakes for this class: 2


# Performance Metrics Evaluation

In [7]:
# Calculate and report classification performance metrics using macro averaging
accuracy = accuracy_score(y_test, preds)
precision = precision_score(y_test, preds, average="macro")
recall = recall_score(y_test, preds, average="macro")
f1 = f1_score(y_test, preds, average="macro")

print("--- Performance Metrics Summary ---")
print(f"Accuracy:  {accuracy:.4f}")
print(f"Precision: {precision:.4f} (Macro)")
print(f"Recall:    {recall:.4f} (Macro)")
print(f"F1-Score:  {f1:.4f} (Macro)")

--- Performance Metrics Summary ---
Accuracy:  0.9694
Precision: 0.9707 (Macro)
Recall:    0.9694 (Macro)
F1-Score:  0.9698 (Macro)



## 1. Executive Summary
In this project, we developed an end-to-end machine learning pipeline to classify handwritten digits using the **Optical Recognition of Handwritten Digits (optdigits)** dataset. Leveraging scikit-learn's `RandomForestClassifier`, we built a robust ensemble model capable of mapping flattened pixel intensity matrices to their corresponding numerical digit classes (0 through 9).

---

## 2. Key Takeaways & Lessons

* **The Power of Ensembles:** Transitioning from individual decision trees to a Random Forest drastically limits variance. By averaging predictions across 150 independent trees, the model achieves strong stability and generalization on unseen test data without requiring deep architecture adjustments.

* **Feature Density Over Complexity:** Although the dataset downscaled the original images into highly compressed 8x8 matrices, the localized spatial patterns of pixel intensities remained dense enough for a classical machine learning algorithm to extract exceptional predictive power.

* **Evaluation Rigor via Macro Metrics:** Relying purely on global accuracy can obscure underlying class blind spots. By evaluating our model with a confusion matrix alongside macro-averaged precision, recall, and F1-scores, we treat every digit class with equal importance, ensuring true operational reliability across all numbers.

* **Continuous Optimization:** Real-world data distributions frequently produce borderline samples (such as handwritten `3`s that closely mimic `8`s). Tracking these programmatic misclassifications allows developers to pinpoint exactly where feature boundaries blur and highlights where targeted hyperparameter tuning is required.